forum/profiles/views.py

```diff
@@ -37,17 +42,76 @@ def perform_create(self, serializer):
     permission_classes = [IsAuthenticated]
-    serializer_class = Serializer
-    queryset = StartupProfile.objects.all()
+    serializer_class = StartupProfileSerializer
+
+    filter_backends = [DjangoFilterBackend, SearchFilter, OrderingFilter]
+
+    search_fields = ['company_name', 'industry', 'country', 'city']
+    filterset_fields = ['industry', 'country', 'city', 'size']
+    ordering_fields = ['company_name', 'created_at']
+
+    def get_queryset(self):
+        """Returns queryset for current user's saved startups"""
+        if self.request.user.is_authenticated:
+            investor = get_object_or_404(InvestorProfile, user=self.request.user)
+            return investor.followed_startups.all()
+
+    def get_serializer_class(self):
+        """Returns the appropriate serializer class based on the action"""
+        if self.action == 'save_startup' or self.action == 'delete_favorite':
+            return Serializer
+        return super().get_serializer_class()
+
+    @swagger_auto_schema(
+        tags=['Save Follow Startups'],
+        manual_parameters=[
+            openapi.Parameter(
+                'industry', openapi.IN_QUERY,
+                description="Filter by industry",
+                type=openapi.TYPE_STRING
+            ),
+            openapi.Parameter(
+                'country', openapi.IN_QUERY,
+                description="Filter by country",
+                type=openapi.TYPE_STRING
+            ),
+            openapi.Parameter(
+                'city', openapi.IN_QUERY,
+                description="Filter by city",
+                type=openapi.TYPE_STRING
+            ),
+            openapi.Parameter(
+                'size', openapi.IN_QUERY,
+                description="Filter by size",
+                type=openapi.TYPE_STRING
+            ),
+        ]
+    )
+    def list(self, request, *args, **kwargs):
+        """Retrieve a list of followed startups"""
+        return super().list(request, *args, **kwargs)
 
-    @action(detail=True, methods=['post'], url_path='save', url_name='save')
-    def post(self, request, pk):
+    @swagger_auto_schema(tags=['Save Follow Startups'])
+    @action(detail=True, methods=['post'], url_path='save-favorite', url_name='save-favorite')
+    def save_startup(self, request, pk):
+        """Add a startup to the user's favourites"""
         investor = get_object_or_404(InvestorProfile, user=request.user)
         startup = get_object_or_404(StartupProfile, pk=pk)
         if startup.followers.filter(pk=investor.pk).exists():
             return Response({'detail': 'Startup is already followed'}, status=status.HTTP_400_BAD_REQUEST)
         startup.followers.add(investor)
         return Response({'detail': 'Startup is saved'}, status=status.HTTP_200_OK)
+
+    @swagger_auto_schema(tags=['Save Follow Startups'])
+    @action(detail=True, methods=['delete'], url_path='delete-favorite', url_name='delete-favorite')
+    def delete_favorite(self, request, pk):
```

Comment:
Add rate-limiting decorators to prevent abuse of the `delete_favorite` actions. Use `django_ratelimit` for custom actions

Reply:
ratelimit decorator added. We're using standard local memory cache which is not optimal for ratelimiting. Ratelimit has to be reconfigured to use Redis when we add it. Note about this added inside settings.py